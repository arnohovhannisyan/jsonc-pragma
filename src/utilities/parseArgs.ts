import { IArguments } from "../models/arguments";

export function parseArgs(match: string | null): IArguments {
	if (!match) return {};

	const entries = [
		...new (typeof URLSearchParams === "undefined"
			? require("url").URLSearchParams
			: URLSearchParams)(match.replace(/\s(\w*?=)/, "&$1"))
	];

	return entries.reduce(
		(acc, [k, v]) => ({
			...acc,
			[k]: v
		}),
		{}
	);
}
