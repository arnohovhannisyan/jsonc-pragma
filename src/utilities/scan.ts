import matchBracket from "match-bracket";
import { ISection } from "../models/section";
import { parseArgs } from "./parseArgs";

function isRegular(line: string) {
	if (/[{[]$/.test(line.trim())) return false;

	return true;
}

/**
 * @param contents The contents of your JSON file
 */
export function scan(contents: string): ISection[] {
	const pragmaRegex = /\s*\/\/\s@(?<name>[a-zA-Z\d]+)(\s(?<args>.*))?$/;

	const lines = contents.split("\n");

	const linesWithPragmas = lines.filter(line => pragmaRegex.test(line));

	return linesWithPragmas.map<ISection>(line => {
		const startingLineNumber = lines.indexOf(line) + 1;
		const startingLine = lines[startingLineNumber];

		const groups = pragmaRegex.exec(line)!.groups!;

		if (isRegular(startingLine)) {
			return {
				start: startingLineNumber,
				end: startingLineNumber,
				args: parseArgs(groups.args),
				name: groups.name
			};
		}

		const startingBracketPos = {
			line: startingLineNumber + 1,
			cursor: startingLine.trimEnd().length
		};

		const endingLineNumber =
			matchBracket(contents.toString(), startingBracketPos).line - 1;

		return {
			start: startingLineNumber,
			end: endingLineNumber,
			args: parseArgs(groups.args),
			name: groups.name
		};
	});
}
