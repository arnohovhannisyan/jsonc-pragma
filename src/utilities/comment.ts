import { ISection } from "~/models";
import { scan } from "~/utilities/scan";

/**
 * @param contents The contents of your JSON file
 * @param selector A function that accepts a section and determines if it should be commented by returning true (comment) or false (don't comment).
 */
export function comment(
	contents: string | Buffer,
	selector: (section: ISection) => boolean = () => true
): string {
	const sections = scan(contents);

	const sectionsToComment = sections.filter(selector);

	const lines = contents.toString().split("\n");

	return lines
		.map((line, i) => {
			if (line.trimStart().startsWith("//")) return line;

			const section = sectionsToComment.find(s => i >= s.start && i <= s.end);

			if (section) {
				const firstLine = lines[section.start];
				const leadingSpaces = /^(\s*)/.exec(firstLine)?.[0].length ?? 0;

				return line.replace(new RegExp(`^(\\s{${leadingSpaces}})`), "$1// ");
			}

			return line;
		})
		.join("\n");
}
