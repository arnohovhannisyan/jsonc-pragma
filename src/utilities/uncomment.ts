import { ISection } from "../models/section";
import { scan } from "./scan";

/**
 * @param contents The contents of your JSON file
 * @param selector A function that accepts a section and determines if it should be uncommented by returning true (uncomment) or false (don't uncomment).
 */
export function uncomment(
	contents: string,
	selector: (section: ISection) => boolean = () => true
): string {
	const sections = scan(contents.toString().replace(/\/\/\s(?!@)/g, ""));

	const sectionsToUncomment = sections.filter(selector);

	const uncommented = contents
		.split("\n")
		.map((line, i) => {
			const shouldUncomment = sectionsToUncomment.some(
				section => i >= section.start && i <= section.end
			);

			if (shouldUncomment) return line.replace("// ", "");

			return line;
		})
		.join("\n");

	return uncommented;
}
