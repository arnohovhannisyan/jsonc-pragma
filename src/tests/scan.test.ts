import { scan } from "../utilities/scan";

test("basic functionality", () => {
	const input = `{
		// @foo bar=5
		"key": true
	}`;

	const output = scan(input);

	expect(output.length).toMatchSnapshot();

	expect(output[0].args).toMatchSnapshot();
	expect(output[0].name).toMatchSnapshot();

	expect(output[0].start).toMatchSnapshot();
	expect(output[0].end).toMatchSnapshot();
});

test("empty pragma", () => {
	const input = `{
		// @ bar=5
		"key": true
	}`;

	const output = scan(input);

	expect(output.length).toMatchSnapshot();
});

test("non-alphanumeric pragma", () => {
	const input = `{
		// @? bar=5
		"key": true
	}`;

	const output = scan(input);

	expect(output.length).toMatchSnapshot();
});

test("numeric pragma", () => {
	const input = `{
		// @1 bar=5
		"key": true
	}`;

	const output = scan(input);

	expect(output.length).toMatchSnapshot();

	expect(output[0].args).toMatchSnapshot();
	expect(output[0].name).toMatchSnapshot();
});

test("multiple arguments", () => {
	const input = `{
		// @foo bar=5 baz=6
		"key": true
	}`;

	const output = scan(input);

	expect(output.length).toMatchSnapshot();

	expect(output[0].args).toMatchSnapshot();
	expect(output[0].name).toMatchSnapshot();

	expect(output[0].start).toMatchSnapshot();
	expect(output[0].end).toMatchSnapshot();
});

test("arguments with spaces", () => {
	const input = `{
		// @foo bar=arg with spaces! baz=6
		"key": true
	}`;

	const output = scan(input);

	expect(output.length).toMatchSnapshot();

	expect(output[0].args).toMatchSnapshot();
	expect(output[0].name).toMatchSnapshot();

	expect(output[0].start).toMatchSnapshot();
	expect(output[0].end).toMatchSnapshot();
});

test("no arguments", () => {
	const input = `{
		// @foo
		"key": true
	}`;

	const output = scan(input);

	expect(output.length).toMatchSnapshot();

	expect(output[0].args).toMatchSnapshot();
	expect(output[0].name).toMatchSnapshot();

	expect(output[0].start).toMatchSnapshot();
	expect(output[0].end).toMatchSnapshot();
});

test("multiple pragmas", () => {
	const input = `{
		// @foo bar=5
		"key": true,

		// @baz bar=10
		"key2": true
	}`;

	const output = scan(input);

	expect(output.length).toMatchSnapshot();

	expect(output[0].args).toMatchSnapshot();
	expect(output[0].name).toMatchSnapshot();

	expect(output[1].args).toMatchSnapshot();
	expect(output[1].name).toMatchSnapshot();

	expect(output[0].start).toMatchSnapshot();
	expect(output[0].end).toMatchSnapshot();

	expect(output[1].start).toMatchSnapshot();
	expect(output[1].end).toMatchSnapshot();
});

test("objects", () => {
	{
		const input = `{
			// @foo bar=5
			"key": {
				"key": true
			}
		}`;

		const output = scan(input);

		expect(output.length).toMatchSnapshot();

		expect(output[0].args).toMatchSnapshot();
		expect(output[0].name).toMatchSnapshot();

		expect(output[0].start).toMatchSnapshot();
		expect(output[0].end).toMatchSnapshot();
	}

	{
		const input = `{
			"key": {
				// @foo bar=5
				"key": true
			}
		}`;

		const output = scan(input);

		expect(output.length).toMatchSnapshot();

		expect(output[0].args).toMatchSnapshot();
		expect(output[0].name).toMatchSnapshot();

		expect(output[0].start).toMatchSnapshot();
		expect(output[0].end).toMatchSnapshot();
	}
});

test("arrays", () => {
	{
		const input = `{
			// @foo bar=5
			"key": [
				{ "key": true }
			]
		}`;

		const output = scan(input);

		expect(output.length).toMatchSnapshot();

		expect(output[0].args).toMatchSnapshot();
		expect(output[0].name).toMatchSnapshot();

		expect(output[0].start).toMatchSnapshot();
		expect(output[0].end).toMatchSnapshot();
	}

	{
		const input = `{
			"key": [
				// @foo bar=5
				{ "key": true }
			]
		}`;

		const output = scan(input);

		expect(output.length).toMatchSnapshot();

		expect(output[0].args).toMatchSnapshot();
		expect(output[0].name).toMatchSnapshot();

		expect(output[0].start).toMatchSnapshot();
		expect(output[0].end).toMatchSnapshot();
	}
});

test("nested objects", () => {
	{
		const input = `{
			// @foo bar=5
			"key": {
				"key": {
					"key": true
				}
			}
		}`;

		const output = scan(input);

		expect(output.length).toMatchSnapshot();

		expect(output[0].args).toMatchSnapshot();
		expect(output[0].name).toMatchSnapshot();

		expect(output[0].start).toMatchSnapshot();
		expect(output[0].end).toMatchSnapshot();
	}

	{
		const input = `{
			"key": {
				// @foo bar=5
				"key": {
					"key": true
				}
			}
		}`;

		const output = scan(input);

		expect(output.length).toMatchSnapshot();

		expect(output[0].args).toMatchSnapshot();
		expect(output[0].name).toMatchSnapshot();

		expect(output[0].start).toMatchSnapshot();
		expect(output[0].end).toMatchSnapshot();
	}
});

test("nested arrays", () => {
	{
		const input = `{
			// @foo bar=5
			"key": [
				[
					{ "key": true }
				]
			]
		}`;

		const output = scan(input);

		expect(output.length).toMatchSnapshot();

		expect(output[0].args).toMatchSnapshot();
		expect(output[0].name).toMatchSnapshot();

		expect(output[0].start).toMatchSnapshot();
		expect(output[0].end).toMatchSnapshot();
	}

	{
		const input = `{
			"key": [
				// @foo bar=5
				[
					{ "key": true }
				]
			]
		}`;

		const output = scan(input);

		expect(output.length).toMatchSnapshot();

		expect(output[0].args).toMatchSnapshot();
		expect(output[0].name).toMatchSnapshot();

		expect(output[0].start).toMatchSnapshot();
		expect(output[0].end).toMatchSnapshot();
	}
});

test("object in array", () => {
	const input = `{
		"key": [
			// @foo bar=5
			{ 
				"key": true
			},
			{ 
				"key": true
			}
		]
	}`;

	const output = scan(input);

	expect(output.length).toMatchSnapshot();

	expect(output[0].args).toMatchSnapshot();
	expect(output[0].name).toMatchSnapshot();

	expect(output[0].start).toMatchSnapshot();
	expect(output[0].end).toMatchSnapshot();
});
