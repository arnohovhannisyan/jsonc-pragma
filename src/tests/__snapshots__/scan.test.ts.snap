// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arguments with spaces 1`] = `1`;

exports[`arguments with spaces 2`] = `
Object {
  "bar": "arg with spaces!",
  "baz": "6",
}
`;

exports[`arguments with spaces 3`] = `"foo"`;

exports[`arguments with spaces 4`] = `2`;

exports[`arguments with spaces 5`] = `2`;

exports[`arrays 1`] = `1`;

exports[`arrays 2`] = `
Object {
  "bar": "5",
}
`;

exports[`arrays 3`] = `"foo"`;

exports[`arrays 4`] = `2`;

exports[`arrays 5`] = `4`;

exports[`arrays 6`] = `1`;

exports[`arrays 7`] = `
Object {
  "bar": "5",
}
`;

exports[`arrays 8`] = `"foo"`;

exports[`arrays 9`] = `3`;

exports[`arrays 10`] = `3`;

exports[`basic functionality 1`] = `1`;

exports[`basic functionality 2`] = `
Object {
  "bar": "5",
}
`;

exports[`basic functionality 3`] = `"foo"`;

exports[`basic functionality 4`] = `2`;

exports[`basic functionality 5`] = `2`;

exports[`empty pragma 1`] = `0`;

exports[`multiple arguments 1`] = `1`;

exports[`multiple arguments 2`] = `
Object {
  "bar": "5",
  "baz": "6",
}
`;

exports[`multiple arguments 3`] = `"foo"`;

exports[`multiple arguments 4`] = `2`;

exports[`multiple arguments 5`] = `2`;

exports[`multiple pragmas 1`] = `2`;

exports[`multiple pragmas 2`] = `
Object {
  "bar": "5",
}
`;

exports[`multiple pragmas 3`] = `"foo"`;

exports[`multiple pragmas 4`] = `
Object {
  "bar": "10",
}
`;

exports[`multiple pragmas 5`] = `"baz"`;

exports[`multiple pragmas 6`] = `2`;

exports[`multiple pragmas 7`] = `2`;

exports[`multiple pragmas 8`] = `5`;

exports[`multiple pragmas 9`] = `5`;

exports[`nested arrays 1`] = `1`;

exports[`nested arrays 2`] = `
Object {
  "bar": "5",
}
`;

exports[`nested arrays 3`] = `"foo"`;

exports[`nested arrays 4`] = `2`;

exports[`nested arrays 5`] = `6`;

exports[`nested arrays 6`] = `1`;

exports[`nested arrays 7`] = `
Object {
  "bar": "5",
}
`;

exports[`nested arrays 8`] = `"foo"`;

exports[`nested arrays 9`] = `3`;

exports[`nested arrays 10`] = `5`;

exports[`nested objects 1`] = `1`;

exports[`nested objects 2`] = `
Object {
  "bar": "5",
}
`;

exports[`nested objects 3`] = `"foo"`;

exports[`nested objects 4`] = `2`;

exports[`nested objects 5`] = `6`;

exports[`nested objects 6`] = `1`;

exports[`nested objects 7`] = `
Object {
  "bar": "5",
}
`;

exports[`nested objects 8`] = `"foo"`;

exports[`nested objects 9`] = `3`;

exports[`nested objects 10`] = `5`;

exports[`no arguments 1`] = `1`;

exports[`no arguments 2`] = `Object {}`;

exports[`no arguments 3`] = `"foo"`;

exports[`no arguments 4`] = `2`;

exports[`no arguments 5`] = `2`;

exports[`non-alphanumeric pragma 1`] = `0`;

exports[`numeric pragma 1`] = `1`;

exports[`numeric pragma 2`] = `
Object {
  "bar": "5",
}
`;

exports[`numeric pragma 3`] = `"1"`;

exports[`object in array 1`] = `1`;

exports[`object in array 2`] = `
Object {
  "bar": "5",
}
`;

exports[`object in array 3`] = `"foo"`;

exports[`object in array 4`] = `3`;

exports[`object in array 5`] = `5`;

exports[`objects 1`] = `1`;

exports[`objects 2`] = `
Object {
  "bar": "5",
}
`;

exports[`objects 3`] = `"foo"`;

exports[`objects 4`] = `2`;

exports[`objects 5`] = `4`;

exports[`objects 6`] = `1`;

exports[`objects 7`] = `
Object {
  "bar": "5",
}
`;

exports[`objects 8`] = `"foo"`;

exports[`objects 9`] = `3`;

exports[`objects 10`] = `3`;
